### PR TITLE
Nano: add transparent forward with ort

### PR DESCRIPTION
**Pull request by TheaperDeng**
```python
# inferencing with `.forward()` (onnx)
model.eval_onnx()
pred = model(x)
```
related to issue #3667 

------------------------------------------------------------------------
After benchmarking, we find that this will give us a 10~20% speed up on ResNet18 with Cifar10 inference latency on i7-9700. (median)

This is a smaller speed up since the onnx forward needs to do 2 extra casts (tensor -> numpy array -> tensor) to support the transparent speed up. Still the influence is not that much.

```python
pl_model.eval()
pytorch_latency = []
with torch.no_grad():
    for i, batch in enumerate(train_loader):
        x, y = batch
        start_time = time.time()
        pl_model(x)
        pytorch_latency.append(time.time() - start_time)
    print("pytorch_latency:", np.median(pytorch_latency)*1000, 'ms')


pl_model.eval_onnx()
onnx_latency = []
with torch.no_grad(): # useless but I leave here for transparent test
    for i, batch in enumerate(train_loader):
        x, y = batch
        start_time = time.time()
        pl_model(x)
        onnx_latency.append(time.time() - start_time)
    print("onnx_latency:", np.median(onnx_latency)*1000, 'ms')
```

### CODE REVIEW

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 1]
```diff
@@ -184,6 +184,53 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    self.exit_onnx()
+
+
+def _forward_onnx(self, *args):
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: What's the difference vs.  `inference`? Should you call `inference` (or have `inference` call  `_forward_onnx`)?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 2]
```diff
@@ -184,6 +184,53 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    self.exit_onnx()
+
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: how to handle `nn.Module.train()`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: I don't think I should add `batchsize` or other additional parameters here for the forward, that is kind of important for some users (e.g. Chronos users) I think and it's only available in `inference`.

Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: The should `inference` call `_forward_onnx`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 2 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: good question, but why would a pl user call `nn.Module.train()`? I will go though some example doc and get back to this comment.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: emmm, but users may directly call `pl_model.inference(x, batch_size=32)` to control the memory usage.

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: This `pl_model(x)` or `pl_model.forward(x)` API is only for those users who want to transform their code to onnxruntime backend with 1 line changes(change `pl_model.eval()` to `pl_model.eval_onnx()`).

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: But may be they can control the batch size them selves by using a data loader like the one I write above. In that case, we may remove the `inference` method.
P.S. I agree `inference` and `forward` should not be provided to users at them same time with such similar functionalities

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 2 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: > good question, but why would a pl user call `nn.Module.train()`? I will go though some example doc and get back to this comment.

It's common practice to call `model.eval` before inference, and call `model.train` before training, no?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: Yes, the user can still call `inference`, but internally `inference` should call `_forward_onnx`, no?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 2 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: For inference yes, since `Trainer.predict` (internally call `.eval()`) is not used frequently. Fore training no, `model.train()` is not required and not used in pl docs and users only needs to call `Trainer.fit`.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:190]: Oh, that's what you mean... I misunderstand that to be using `_forward_onnx` replace `inference`. Yes, I will change it. THX

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 2 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: Not for PL; but it should be used for standard PyTorch code?

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]:
```python
pl_model = Trainer.compile(model, loss, optim, onnx=True)
```
Users will get a pl model, and users will use it as a pytorch model?

Emm, it's possible and (maybe) valid. I will check some details later.

Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: > ```python
> pl_model = Trainer.compile(model, loss, optim, onnx=True)
> ```
> 
> Users will get a pl model, and users will use it as a pytorch model?
> 
> Emm, it's possible and (maybe) valid. I will check some details later.

The goal is to transparently accelerate both PL and PyTorch code

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: Surprisingly, it works well... (this example is from https://github.com/pytorch/examples/blob/master/mnist/main.py)
Only 4 lines to change and we can use the training speed up in bigdl-nano-init and inference speed up in onnxruntime
https://github.com/TheaperDeng/git_learning/compare/mnist-example-nano

Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: > Surprisingly, it works well... (this example is from [pytorch/examples@`master`/mnist/main.py](https://github.com/pytorch/examples/blob/master/mnist/main.py?rgh-link-date=2021-12-14T08%3A58%3A34Z)) Only 4 lines to change and we can use the training speed up in bigdl-nano-init and inference speed up in onnxruntime [TheaperDeng/git_learning@`mnist-example-nano` (compare)](https://github.com/TheaperDeng/git_learning/compare/mnist-example-nano)

What is `model.eval_onnx(tuple(next(iter(test_loader))[:-1]))`?

I think the below code might be problematic:

```python
model = Trainer.compile(model, nn.NLLLoss(), optimizer, onnx=True)

model.eval_onnx()
...

#model.exit_onnx() is not called
model.train()
...
```

And if the user just wants to use ONNX, it does not make sense to ask the user to specify loss and optimizer in `compile` (the user may not know what loss or optimizer is at all if he or she does not train the model).

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:188]: 1. If users never call `Trainer.fit(model, data)` they need to put an example input in `eval_onnx`. I can accept a dataloader to be the input later, that is `model.eval_onnx(test_loader)`
2. Yes, I will check and find a way to make it work later...
3. We may allow some thing like this
```python
# model is a nn.module
model = Trainer.compile(model, onnx=True)
# use model like a nn.module
```

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 3]
```diff
@@ -188,6 +188,11 @@ def on_fit_start(self):
 
 
 def _forward_onnx(self, *args):
+    # if self.train() is called without calling self.exit_onnx()
+    if self.training:
+        self.exit_onnx()
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:193]: shall we set `self._ortsess_up_to_date = False` here?

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:193]: yes, and since we need to support pytorch users other than pl users, there are many details to be considered. I will switch this PR to WIP for now.

Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:193]: > yes, and since we need to support pytorch users other than pl users, there are many details to be considered. I will switch this PR to WIP for now.

I think you should just override the `train` method?

**Comment by TheaperDeng** [comment]
```python
# Pytorch-lightning users
model = Trainer.compile(model, loss=loss, optimizer=optimizer, onnx=True)

trainer = Trainer()
trainer.fit(model, data)

model.inference(x, batch_size=32)  # ort backend
model.inference(x, batch_size=32, backend=None)  # pytorch backend

model(x)  # pytorch backend
model.eval_onnx()
model(x)  # ort backend

trainer.fit(model, data)  # incremental fitting
```
```python
# pytorch users
model = Trainer.compile(model, onnx=True)

# >>>> start their own training loop >>>>
model.train()
# ...
# <<<< end their own training loop <<<<

model.inference(x, batch_size=32)  # ort backend
model.inference(x, batch_size=32, backend=None)  # pytorch backend

model(x)  # pytorch backend
model.eval_onnx(input_sample)
model(x)  # ort backend

# >>>> start their incremental training loop >>>>
model.train()
# ...
# <<<< end their incremental training loop <<<<
```

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 4]
```diff
@@ -184,6 +188,66 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    self.exit_onnx()
+
+
+def train(self, mode=True):
+    self.exit_onnx()
+    self._train_mode(mode)
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:196]: why add a new `_train_mode()` instead of calling `super().train()`?

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:196]: `super()` need a class name when it is defined outside a class. And I can't get it here.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 5]
```diff
@@ -184,6 +188,66 @@ def inference(self,
...
+    # change to eval mode
+    self.eval()
+
+    # get input_sample
+    if isinstance(input_sample, DataLoader):
+        input_sample = tuple(next(iter(input_sample))[:-1])
+    if input_sample is None and self.example_input_array:
+        input_sample = self.example_input_array
+    if input_sample is None and self.trainer is None:
+        raise RuntimeError("You must state an input_sample or fit on the model to use `eval_onnx`.")
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:232]: You must specify an input_sample or call Trainer.fit on the model first to use `eval_onnx`

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 6]
```diff
@@ -184,6 +188,66 @@ def inference(self,
...
+    :param sess_options: (optional) ortsess options in ort.SessionOptions type.
+    :param **kwargs: (optional) will be passed to torch.onnx.export function.
+    '''
+    # change to eval mode
+    self.eval()
+
+    # get input_sample
+    if isinstance(input_sample, DataLoader):
+        input_sample = tuple(next(iter(input_sample))[:-1])
+    if input_sample is None and self.example_input_array:
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:229]: elif

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 7]
```diff
@@ -208,5 +272,11 @@ def bind_onnxrt_methods(pl_model: LightningModule):
     pl_model.update_ortsess = partial(update_ortsess, pl_model)
     pl_model.on_fit_start = partial(on_fit_start, pl_model)
     pl_model.inference = partial(inference, pl_model)
+    pl_model.eval_onnx = partial(eval_onnx, pl_model)
+    pl_model._forward_onnx = partial(_forward_onnx, pl_model)
+    pl_model.exit_onnx = partial(exit_onnx, pl_model)
+    pl_model._torch_forward = pl_model.forward
+    pl_model._train_mode = pl_model.train
+    pl_model.train = partial(train, pl_model)
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:281]: for `_forward_args`, you may explicitly check if the model is `LightningModuleFromTorch`; if not, `pl_model.model` will be incorrect.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 6 above)
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:229]: Will `example_input_array` be set somewhere?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 4 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:196]: `self._ortsess_up_to_date = False`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:199]: So `_forward_onnx` will always return numpy? Is it compactible with `forward`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 8]
```diff
@@ -184,6 +188,66 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    self.exit_onnx()
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:191]: why don't you call the original `on_fit_start` here, just like `train`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 6 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:229]: yes, it's a property in pytorch lightning. Users may set it themselves

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 9]
```diff
@@ -29,6 +32,8 @@
                             '_build_ortsess',
                             'update_ortsess',
                             'on_fit_start',
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:34]: remove 'on_fit_start' from the list?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 10]
```diff
@@ -184,6 +188,68 @@ def inference(self,
 # on_fit_start (LightningModule method overwrite)
 def on_fit_start(self):
     self._ortsess_up_to_date = False
+    return self._on_fit_start_old()
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:191]: also `self.exit_onnx()`?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:200]: And does conversion between numpy and tensor impact latency?

**Comment by jason-dai** [comment]
See comments inline; otherwise LGTM

**Comment by jason-dai** [comment]
BTW, you also need support the follow in a separate PR:
1) `loss` and `optimizer` should be optional in `compile` for ONNX and quantization support
2) Handle the interaction between ONNX and quantization

**Comment by TheaperDeng** [comment]
> BTW, you also need support the follow in a separate PR:
> 
> 1. `loss` and `optimizer` should be optional in `compile` for ONNX and quantization support
> 2. Handle the interaction between ONNX and quantization
> 3. Add `export_onnx` support?

for 1,2 I will have some look. For 3, I think `to_onnx` is always there(a built-in method in pl) for the compiled model.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 1 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:200]: not really, I did some tests and the impact is minor

**Comment by TheaperDeng** [comment]
http://10.112.231.51:18888/job/ZOO-PR-NanoTests/147/
http://10.112.231.51:18888/job/BigDL-PRVN-chronos-Python-Spark-2.4-py37-ray-part1/173/
